### ISSUE TITLE
DELIMITER on MySQL is not understood as expected

### ISSUE THREAD

**Issue by niconoe-** [sqlfluff/sqlfluff#3599]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

On MySQL, it is possible to run the statement `DELIMITER` to specify a new delimiter that separates the next statements. This is useful for defining procedures for instance where a single statement (create the procedure) must contain several statements (statements within the procedure).

But even outside a procedure, such `DELIMITER` statement is not taken into account by SQLFluff.

### Expected Behaviour

Try this [online](https://online.sqlfluff.com/) with "mysql" dialect:

```sql
DELIMITER $$

SELECT 1 $$

DELIMITER ;
```

This is valid MySQL snippet that should not trigger any SQLFluff error about the character that must terminate a statement.

### Observed Behaviour

Nevertheless, the observed behavior is the following: https://online.sqlfluff.com/fluffed?sql=H4sIAFoVzWIC_3Nx9fH09QxxDVJQUeHl4uUKdvVxdQ5RMITwXOCy1gD_88-jKQAAAA%3D%3D&dialect=mysql

```
PRS | 1 / 10 | Line 1, Position 10: Found unparsable section: ' $  SELECT 1$  DELIMITER'
L052 | 5 / 10 | Statements must end with a semi-colon.
```

Because of the 1st delimiter section not ending with a semi-column, the content analysed for parsing is the whole snippet (until the 1st semi-column found), which is not the correct way to parse such MySQL statement. Therefore, no specific delimiter is understood in the context of SQLFluff, and this causes the L052 error.

### How to reproduce

To reproduce, please try the following snippet in mysql dialect:

```sql
DELIMITER $$

SELECT 1$$

DELIMITER ;
```

### Dialect

mysql

### Version

I'm using the latest version of SQLFluff: currently set to 1.1.0 (when I wrote this issue).

### Configuration

== Raw Config:
core:                                   
    dialect:            mysql               
    disable_noqa:       False               
    encoding:           utf-8               
    exclude_rules:      L042                
    fix_even_unparsable:False               
    ignore:             ['parsing']         
    ignore_templated_areas:True                
    nocolor:            False               
    output_line_length: 80                  
    recurse:            True                
    rules:              all                 
    runaway_limit:      10                  
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          raw                 
    verbose:            2                   
indentation:                            
    indented_ctes:      False               
    indented_joins:     True                
    indented_on_contents:True                
    indented_using_on:  True                
    template_blocks_indent:True                
rules:                                  
    allow_scalar:       True                
    comma_style:        trailing            
    indent_unit:        space               
    max_line_length:    120                 
    single_table_references:consistent          
    tab_space_size:     4                   
    unquoted_identifiers_policy:all                 
    L003:                                   
        hanging_indents:    True                
    L007:                                   
        operator_new_lines: after               
    L010:                                   
        capitalisation_policy:upper               
    L011:                                   
        aliasing:           explicit            
    L012:                                   
        aliasing:           explicit            
    L014:                                   
        extended_capitalisation_policy:upper               
    L016:                                   
        ignore_comment_clauses:False               
        ignore_comment_lines:True                
    L026:                                   
        force_enable:       False               
    L027:                                   
    L028:                                   
        force_enable:       False               
    L029:                                   
        unquoted_identifiers_policy:all                 
    L030:                                   
        capitalisation_policy:upper               
        extended_capitalisation_policy:consistent          
    L031:                                   
        force_enable:       False               
    L036:                                   
        wildcard_policy:    single              
    L038:                                   
        select_clause_trailing_comma:forbid              
    L040:                                   
        capitalisation_policy:upper               
    L042:                                   
        forbid_subquery_in: join                
    L047:                                   
        prefer_count_0:     False               
        prefer_count_1:     True                
    L051:                                   
        fully_qualify_join_types:inner               
    L052:                                   
        multiline_newline:  False               
        require_final_semicolon:True                
    L054:                                   
        group_by_and_order_by_style:consistent          
    L057:                                   
        additional_allowed_characters:""                  
        allow_space_in_identifier:False               
        quoted_identifiers_policy:all                 
        unquoted_identifiers_policy:all                 
    L059:                                   
        force_enable:       False               
        prefer_quoted_identifiers:True                
    L062:                                   
    L063:                                   
        extended_capitalisation_policy:upper               
    L064:                                   
        force_enable:       False               
        preferred_quoted_literal_style:single_quotes       
    L066:                                   
templater:                              
    unwrap_wrapped_queries:True                
    jinja:                                  
        apply_dbt_builtins: True                

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
I think this is going to be a tough one to fix. The way the parser works has the delimiter hardcoded at present and not sure how easy that would be to dynamically update mid-query (and then revert back for next query?).

**Comment by cphill02** [comment]
Is there any workaround for this? (v  1.4.2 )

This issue makes using sqlFluff unusable for most every MySQL Stored Procedures. Most redefine the ; delimiter.

**Comment by taticaramos** [comment]
does anyone found workaround for this?

**Comment by theking2** [comment]
The token for statement delimiter is dynamic. I believe this is the only computer langauge where semantics (parsing) can change the tokenizer (lexing). But this should not be impossible to code. In which source is the tokenizer programmed?

**Comment by fdobad** [comment]
same here, using vim and ALELinter: `W 5   DELIMITER // /* W: Line 5, Position 11: Found unparsable section: '//`


<details>
    # sqlfluff, version 3.4.2
  
    [sqlfluff]
    dialect = mariadb
    
    [sqlfluff:indentation]
    indent_unit = space
    tab_space_size = 2
</details>

**Comment by theking2** [comment]
Does using`##` as delimiter help?

**Comment by Mynk11** [comment]
I built a working prototype for this, and it does work — but it runs into a wall that I think needs a maintainer decision before it's worth turning into a PR. Posting the findings rather than the patch.

### It's tractable on the Python lexer

@tunetheweb's read in 2022 was that the delimiter is hardcoded and hard to change mid-file. The delimiter is hardcoded, but the lexer turns out to be a reasonable place to fix it: `PyLexer.lex_match` consults its matchers at every token boundary, so a matcher that carries state can track the active delimiter itself. No changes to `lex_match`, the parser, or the file grammar are needed.

The prototype is ~130 lines across two files:

- **`DelimiterCommandLexer`** (in `dialect_mysql.py`), a `StringLexer` subclass which matches three things: the `DELIMITER` keyword, its argument (recording it as the new delimiter), and any later occurrence of the active delimiter, emitted as a `statement_terminator`.
- **One core hook.** Matchers are shared across the whole dialect, so a stateful one has to be per-file. I added `StringLexer.fresh()` (returns `self` by default) and had `PyLexer.__init__` build its list via `[m.fresh() for m in ...]`. Stateless matchers are unaffected.
- **`DelimiterGrammar`** gains a `statement_terminator`-typed alternative. Notably mysql already special-cases one non-`;` delimiter here — `Ref("TildeSegment")` — so the shape isn't new.

The argument to `DELIMITER` is lexed as the terminator of that command rather than as a separate node, which means `DelimiterStatement` stays exactly as it is today (`Ref.keyword("DELIMITER")`) and `Delimited()` splits the file correctly with no grammar change.

Results — all of these parse cleanly, where every one is unparsable on `main`:

```sql
DELIMITER $$
SELECT 1 $$
DELIMITER ;
```
```sql
DELIMITER //
CREATE PROCEDURE p() BEGIN SELECT 1; END //
DELIMITER ;
```
```sql
DELIMITER $$
CREATE TRIGGER t BEFORE INSERT ON x FOR EACH ROW BEGIN SET @a = 1; END$$
DELIMITER ;
```

That third one is #5862. `;` inside the procedure body still lexes as an ordinary `semicolon`, so the body splits correctly. `DELIMITER` used as an identifier (`SELECT delimiter FROM t`) is unaffected — the keyword only matches when followed by a token starting with a non-word character, which a delimiter always is.

Test results: **6933 dialect tests pass**, and **2813 rules + CLI tests pass**. The 3 failures I see in `test/core/plugin_test.py` are pre-existing in my environment (example plugin not installed) and identical with the change stashed. `black` and `ruff` clean.

### The wall: there are two lexers

This is the part I can't decide unilaterally.

1. **The Rust lexer bypasses it entirely.** `get_lexer_class()` returns `PyRsLexer` whenever `sqlfluffrs` is available, so anyone on `sqlfluff[rs]` would get today's behaviour with no indication anything was meant to be different. Silently working for some users and not others seems worse than not shipping it.

2. **It breaks the codegen.** `utils/build_lexers.py` does `lexer_class = lexer_matcher.__class__.__name__` and raises `ValueError` on anything that isn't `StringLexer` or `RegexLexer`. A third matcher class needs a Rust counterpart, and a stateful one specifically needs the Rust lexer to support per-file matcher state.

So the Python half is maybe a day's work and the Rust half is the actual project. Which direction would you want?

- **(a)** Land the Python side behind something explicit, and treat Rust parity as a follow-up — accepting a temporary behaviour split between the two engines.
- **(b)** Hold until it can be done in both, and use the prototype as the spec for the Rust side.
- **(c)** Something else entirely — e.g. handling `DELIMITER` as a source pre-processing step, which would sit above both lexers but needs source-position mapping to survive `fix`.

Happy to write up whichever, or to just leave the prototype here if it's useful to whoever picks it up. I didn't want to drop a core-lexer change on you unannounced given `CONTRIBUTING.md` asks for context and trade-offs on architectural PRs.

*Disclosure per the AI-assisted contributions section of CONTRIBUTING.md: this was AI-assisted (Claude). The prototype, the test runs, and both blockers above were executed and verified against a local checkout at 6166dcc, not inferred — including confirming `_HAS_RUST_LEXER` is what selects the engine.*

**Comment by rfanth** [comment]
Just to clarify, #8531 DOES NOT fix this problem and can't because the lexer is the issue, at the `$$`, before any grammar is consulted, but without #8531, fixing this isn't going to get you very far because the `BEGIN... END` structure with more than one statement is fundamentally broken pre #8531 as the scope currently fails after the first statement without it.

The lines you have above are minimal reproductions with only one statement within the BEGIN... END, but I'm sure you looked at it to start with because you wanted to do something more extensive with it.

**Comment by rfanth** [comment]
As a workaround, `DELIMITER ;;` instead of `DELIMITER $$` works pretty well to avoid the lexer issue only.  `##` doesn't work, sadly, MySQL treats it as a comment which makes using it as a lexing bypass inadvisable.
